### PR TITLE
Add --url, --url-only and load-options

### PR DESCRIPTION
**Pull request by coltnz**
**Status**
First take implementation with pretty extensive tests. We've started using it at SMX. 

**Changes**
New switches -u and --url <url> added to zprint.main

`load-options!` takes url, loads content as edn to pass to set-options!

Content is cached in new `~/.zprint/urlcache dir`, respecting
Cache-Control max-age, the Expires header, or failing those, the 5 min default.

If the url has bad content or is unavailable, an expired cache will be used but a warning will
be logging to System.err, else an Exception thrown.

**Comments / Queries**
I decided if URL didn't work to fall back to an expired cache.  This necessitated logging a warning to System.err

`*default-url-cache-secs*` set to 300secs? Seemed a reasonable tradeoff. 

Resisted adding --url-no-cache (delete cache/wait 5 mins) or --url-strict (delete cache)

Etags felt a bit overkill.

### CODE REVIEW

File: src/zprint/main.cljc [hunk 1]
```diff
@@ -80,13 +81,18 @@
...
+
         [option-status option-stderr switch?]
           (if (and (not (clojure.string/blank? options))
                    (clojure.string/starts-with? options "-"))
             ; standard not yet implemented
-            (if (or version? help? default? #_standard? explain?)
-              [0 nil true]
-              [1 (str "Unrecognized switch: '" options "'" "\n" help-str) true])
+            (cond (or version? help? default? #_standard? explain?) [0 nil true]
+                  url? (try (load-options! (second args))
```
Comment (coltnz) [inline src/zprint/main.cljc]: missing #?(:clj  here

**Comment by kkinnear** [comment]
Thanks for this! I'm chasing a bug with the graalVM binaries, so it will be a bit before I can give this a good look.  I'll get to it soon.

**Comment by coltnz** [comment]
FWIW: I built graal on top of this patch and seems to work fine with:

 /Library/Java/JavaVirtualMachines/graalvm-ce-19.2.1/Contents/Home/bin/native-image -H:EnableURLProtocols=https,http  -H:+ReportExceptionStackTraces  --initialize-at-build-time  --report-unsupported-elements-at-runtime -jar target/zprint-filter-0.5.1

Didn't go through all the switches though.

File: src/zprint/main.cljc (hunk 1 above)
Comment (coltnz) [reply src/zprint/main.cljc]: added

**Comment by kkinnear** [comment]
Nice work on trying out graalVM.  I would have spent a good while before I figured out `-H:EnableURLProtocols=https,http`, assuming that I ever figured it out.  Integrating this is next up on my list.

**Comment by kkinnear** [comment]
I'm integrating this now.  An impressive piece of work, I'm learning things I didn't know about both Clojure and the Java libraries!

I expect to have more questions, but one stands out now:  What is your thinking on `$HOME/.zprintrc` configuration information?  I haven't tested it yet, but from the code it looks to me like the call to load-options will pick up the configuration information in the `$HOME/.zprintrc` file before it applies the information from the URL.  

1. Is that what you intended?
2. Is that how you are using it in your environment?  People can set stuff in their `$HOME/.zprintrc` files and you then override whatever you want with the configuration from the URL?  (Which is what I think is happening now, but as I said, I haven't actually tried it yet.)

Thanks!

**Comment by kkinnear** [comment]
I have now tested it, and yes, the `$HOME/.zprintrc` file is used.  Which isn't bad, I'm just trying to figure out what you had in mind.  I think that some people would want it to not be used, and most people would want it to be used.  It would be fairly straightforward to add another switch (which would be hard to name) which would take a URL and __not__ include the `$HOME/.zprintrc` file if someone wanted that at some point.

**Comment by coltnz** [comment]
Yeah sorry should have that made clear. My thoughts at the time was we wouldnt use it but best to not violate least surprise for existing users. It now occurs to me however that my devs might still use zprint in other cases and I would quite like the protection of a strict mode.  ```--url-only```  would be easy enough to add, wdyt?

**Comment by kkinnear** [comment]
Yes, it might be good add `--url-only`.  I don't see an obvious single letter switch, and two letter switches are ok but kind of odd.  Maybe we just do `--url-only`?  But I'm open to suggestions here.

**Comment by kkinnear** [comment]
Of course, I was thinking about letting people configure the cache-timeout and cache-path in their .zprintrc file, and if we ignore it, then that doesn't work well.  One kind of odd approach would be to examine that .zprintrc file for these values and use them even for `--url-only`, but not use anything else from .zprintrc.  That makes it all a bit harder to implement, but it might be nice to be able to configure that.  Or not.  I obviously think configuring just about everything is a good idea, but I'm also open to not configuring this stuff.

**Comment by coltnz** [comment]
Yep seems reasonable, I'll see what I can do.

**Comment by kkinnear** [comment]
There is a recent issue which involves configuration about configuration, kind of like putting the cache-timeout and location in the options map even when doing `--url-only`.  I'm going to try to solve both problems with some kind of pre-configuration `.zprintrc` scan -- so that the cache-timeout and location could be found from some `.zprintrc` even when using `--url-only` (and certainly when using `--url`. 

I haven't done this yet, but I'm telling you before I work on it so that (hopefully) I don't invalidate any work you were doing in that direction.  I've worked in groups where I said I was going to do something and then someone else would leap into doing it while I was working on it and I __really__ didn't like that.  So I'm trying to avoid doing that with you.  Feel free to implement `--url-only` (or not, I really don't mind just adding it myself), but don't put any (additional?) effort into figuring out how to get the cache-timeout and location into the configuration.  I'll do that part along with solving this other issue.  

If you have already done something, let me know and we'll see where we go from there.

Sorry to (possibly) step on your toes here.

**Comment by coltnz** [comment]
np, ill push once done bar config

**Comment by coltnz** [comment]
So I think this far as I can go.  Uses :url-cache-path or :url-cache-secs  if there but relies on             (get-options) atm which doesnt provide the keys yet.

**Comment by coltnz** [comment]
So I followed my nose and had a stab at this. 

I added: 
```                                                (when url-only? (set-options! {:formatting-configured? true}))```

which eventuates in  (defn config-set-options!):

```  
...
(when (and (not (:configured? (get-options)))
              (not (:configured? new-options)))
     (if (:formatting-configured?)
       (config-configure-non-formatting!)
       (config-configure-all!)))
```

From there seems we require a list of formatting or non-formatting keys to be maintained.  I guess the non formatting list might be short & maintainable? Or is there some other heuristic we can apply?

**Comment by coltnz** [comment]
formatting-configured should of course be false

**Comment by kkinnear** [comment]
Thanks.  I apologize for not corresponding sooner on this. I actually have the code for this completely finished, both `--url` and `--url-only`.  It took a bit of reorganization to figure out how to pre-configure enough to figure how to configure things.  I ended up doing something more or less like what you have suggested.  

I'm sorry that I haven't yet released the result -- in the process of documenting the various changes for this release, I decided to overhaul the documentation (i.e., the `README.md` and everything else).  I'm adding a host of task oriented short `.md` files, and will keep the current readme as "reference".  I am probably two weeks away from finishing with the docs and releasing the new version (not my favorite activity by any means, but I'm pushing through it).

We have a couple of ways to go from here.   Presently I'm operating on a branch, and have not rebased the work back onto the main branch.
  1. I can continue on with the docs, and release it when I get it all documented and merged and tested.
  2. I could push this branch to github and you could try it out see if it does what you think it should do.
  3. Something else ...

I have created a set of tests for `--url` and `--url-only` that operate on the actual uberjar, not the testing framework.  That's in addition to the tests you created for the expectations testing framework (which I extended to work with the changes that I made to basic zprint).

In the event you want to experiment (test) what I have already, the new test script which will test either the uberjar or graalVM binaries directly is called `test_config`.  You run it by giving it the version and whether it is the uberjar or graalVM binaries:
```
Kims-MacBook-Air:zprint kkinnear$ ./test_config 0.5.4 uberjar
Saving ~/.zprintrc as ~/.zprintrc.save
If you ^c out of this test, be sure and clean up ~/.zprintrc and ./.zprintrc!
...........  --url
...........  --url, with URL missing, see if cache works
...........  --url, with URL missing, see if expired cache works
...........  --url, with URL missing, and no cache
...........  --url, with URL not a valid Clojure map, and no cache
...........  --url, with URL not a valid zprint options map, and no cache
...........  --url-only, with valid URL and no cache
...........  --url-only, with valid URL, invalid file w/fn, and no cache
...........  basic config test.
...........  {:cwd-zprintrc? true} on command line test
...........  {:search-config? true} on command line test
...........  ~/.zprintrc being used test
...........  -d test -- ~/.zprintrc should not be used
...........  bad ~/.zprintrc, invalid Clojure map test
...........  invalid zprint options map in ~/.zprintrc
...........  -v test
...........  -h test
...........  -e test
...........  :coerce-to-false test
...........  fn in ~/.zprintrc file test
...........  fn is invalid in ./.zprintrc file test
...........  fn on command line test
Restoring ~/.zprintrc
Kims-MacBook-Air:zprint kkinnear$ 
```
It spits out the tests it runs and any errors it finds.  

This test script is quite the hack, in that the way that it sees if things "work" is (in many cases) by counting lines.  Which is pretty fragile (e.g., when I change the options map, this breaks).  But it does serve the purpose better than what I had before, which was nothing.  In the event that you try this script eventually (either now or after I release it), I would be interested in your take on whether or not it is actually testing what it says it is testing for the `--url` and `--url-only` options.

Anyway, let me know which way you want to go here.  And again, my apologies for getting sidetracked on the documentation.

Thanks!

**Comment by coltnz** [comment]
Thanks for the update. If you can push a branch I'm happy to help test and verify.

**Comment by kkinnear** [comment]
Great!  I pushed branch `urlcfg`.  This is the branch I'm working on to finish the documentation (it has a bunch of new, very incomplete docs in the doc directory).  It also has all of the code that I've integrated for the next release, including `--url` and `--url-only`.  It should also have `test_config` in the top level directory.  I'm interested in what you make of it all.  

 I've done a lot with the uberjar, and I've built the graalVM binary for macOS a couple of times (after including your changes in the build script, of course), and it worked.  Though I've done a bit of work since I last tried the graalVM build.

If you build the graalVM binary, you can use `test_config` to run it -- just say: `./test_config 0.5.4 graalvm-mac` or `graalvm-linux` if you use Linux.

Thanks for anything you can try with this!

**Comment by coltnz** [comment]
The `(load-string)` in `(get-config-from-file)` is going to be problematic for Graal, no?

**Comment by kkinnear** [comment]
Sigh.  Yes, you are so right.  Given that nobody has ever complained about not being able to use functions in options maps, I guess I will back out this change and leave it for another day. 

About the only possibility I see would be to use [sci](https://github.com/borkdude/sci), which is supposed to work with graalVM, and provide a protected environment for execution as well -- which would be a plus.  This would allow functions in options maps anywhere.   But I'm not wild about adding dependencies to zprint.  

One alternative would be to let the uberjar do the `load-string` and force the graalVm binaries to do `read-string`, assuming that I can figure out how to determine that something was built using the graalVM `native-image` command.  Which I have not yet done.  There must be some way to figure this out at run-time, but I haven't found it yet.

More to do before releasing.  Thanks for alerting me to this.

**Comment by coltnz** [comment]
This seems to work:

```
load-fn   (if (and (nil? (System/getProperty "org.graalvm.nativeimage.imagecode"))
                                                acceptfns?)
                                         load-string clojure.edn/read-string)
opts-file (load-fn (apply str lines))]
```

Ad-hoc tests:

```
zprint git:(urlcfg) ✗ echo "{:width 500}" > ~/.zprintrc; java -jar  ./target/zprint-filter-0.5.4 < ./test.clj |wc -l
      81
zprint git:(urlcfg) ✗ echo "{:width (inc 499)}" > ~/.zprintrc; java -jar  ./target/zprint-filter-0.5.4 < ./test.clj |wc -l
      81
zprint git:(urlcfg) ✗ echo "{:width 500}" > ~/.zprintrc; ./zprint-filter-0.5.4  < ./test.clj |wc -l
      81
zprint git:(urlcfg) ✗ echo "{:width (inc 499)}" > ~/.zprintrc; ./zprint-filter-0.5.4  < ./test.clj |wc -l
In Home directory file: /Users/colint/.zprintrc, The value of the key-sequence [:width] -> (inc 499) was not a number
```

Notes:

The Graal selector is pretty broad, there maybe specific Graal modes which use stubs that could support this <shrug>.  Pulling out the property as `config/graal?` might be  generally useful too.

**Comment by kkinnear** [comment]
Thanks!  That was a real help.  Ultimately, I moved in a slightly different direction -- the code now wraps the use of `load-string` with a `try-catch`, and if it fails (for either of two explicit reasons), the code falls back to `read-string`.  That way, I sidestep the problem of recognizing graalVM at runtime, and if it ever starts working, it will just work.  Of course, I added the unpleasantness of recognizing two explicit problems that graalVM detects.  I still haven't tried it with Linux yet.

I've done as much documentation reorganizing as I am going to do for now.  I toyed with moving the old `README.md` (which is now `doc/reference.md`) into `.adoc` format so that I could get an automatic table of contents, and I could get nicer default value text formatting, but it would be a week or even weeks before I finished that.  I've decided to leave that for later, and move into the release and testing cycle now.  I spent a good bit of time making sure that cljdoc works with the new documentation approach -- always a challenge, since the analysis phase frequently runs into problems that don't show up anywhere else.  But worth doing, as cljdoc is not only a great tool but use `zprint` as well!

Which is why I'm telling you all of this -- the `urlcfg` branch should be working for graalVM now (at least it did when last I tested it).  I'm going to back out my `.adoc` trials, and move into some testing on that branch, then I'll rebase onto master, do a bit more testing, and then release it all.  Which usually turns into some days of work, but the `test_config` script should help.  At least it will give me more confidence, if not speed things up.

If you have found anything that doesn't work, this would be the time to let me know.  

I hope that you are not impacted by the pandemic excitement.  Looks like NZ is largely virus free, from what I can see.  In contrast to where I am, where it is just really starting to show up.   All of our schools just closed.  God forbid you should want to buy toilet paper!  Going to be an interesting time.

**Comment by kkinnear** [comment]
Thanks for the code for this.  I manually merged the changes, but fortunately GitHub managed to figure out you are a contributor.  I would have fumbled around a good bit trying to do this.  I was particularly impressed with the cacheing you did.  I did rename all of the cache location things (and made them configurable, even for `--url-only`).  This is a neat capability and it was great working with you!  Thanks again!

**Comment by coltnz** [comment]
Thanks for helping me get this in. I think its going to be useful addition for teams to have standardised Clojure formatting at last. Though I still need to work out what do with Graal and our internal CA. 

Full lockdown here now so we're not immune in any sense. Good luck!

**Pull request closed without merging**